### PR TITLE
dns: serve localhost names directly and retry loopback-only answers without AI_ADDRCONFIG

### PR DESCRIPTION
**Pull request by robobun**
### Problem

On a Linux host whose only IPv6 address is `::1` — every container on an IPv4-only Docker network — nothing in Bun can reach a `Bun.serve()` / `Bun.listen()` bound to `"localhost"` on the same machine. `fetch()`, `new WebSocket()`, `Bun.connect()` and `bun install --registry http://localhost:PORT/` all fail with `ConnectionRefused` / `ECONNREFUSED` / close code 1006.

The two halves of Bun disagree about what `localhost` means:

- **Listen side** (`bsd_create_listen_socket`, `packages/bun-usockets/src/bsd.c`) resolves with `AI_PASSIVE` only and walks `AF_INET6` first, so the listener sits on `[::1]`.
- **Connect side** (`internal::work_pool_callback`, `src/runtime/dns_jsc/dns.rs`) resolves with `AI_ADDRCONFIG`. glibc decides whether a family is "configured" from non-loopback interfaces only, so on such a host `getaddrinfo("localhost")` returns `127.0.0.1` (twice) and even `getaddrinfo("ip6-localhost")` — a hosts-file name for `::1` only — returns `127.0.0.1`. usockets is never given `::1` to try.

In-repo reproducers on such a host with 1.4.0: `fetch-preconnect.test.ts` (11/14 fail), `bun-serve-html-entry.test.ts` (5/7), bun-install "should support --registry CLI flag", and the `localhost` tests in `socket.test.ts` / `bun-connect-x509.test.ts`. #36543 and #35160 worked around it test by test. Node 26 has the identical bug (output below).

### Fix

`default_hints` was copied from Chromium's resolver, but the two loopback rules Chromium pairs with `AI_ADDRCONFIG` were never ported. This PR adds both to the connect-path resolver:

1. **`localhost` and `*.localhost` never reach the system resolver.** `is_localhost_name` is Chrome's `IsLocalHostname` (ASCII case-insensitive, one trailing dot stripped); `internal::getaddrinfo` completes those names synchronously with `[::1, 127.0.0.1]` (`localhost_results`, Chrome's `ServeLocalhost`, RFC 6761 §6.3). The entry goes through the normal cache and `after_result_entries`, so usockets sees an ordinary completed lookup, and it is narrowed by `BUN_FEATURE_FLAG_DISABLE_IPV4` / `_IPV6` like any other.
2. **Loopback-only answers are retried without `AI_ADDRCONFIG`.** The existing `EAI_NONAME` retry in `work_pool_callback` now also fires when the call succeeded but every address is loopback and only one family is present (`is_all_loopback_of_one_family`, Chrome's `IsAllLocalhostOfOneFamily`, crbug 42058 / 49024). This is what fixes `ip6-localhost`, `ip6-loopback`, `localhost6` and custom hosts-file loopback names. If the retry fails, the first answer is kept.

Why connect-side: the connect path already races every address it is given (up to four at once), so handing it both loopback addresses costs at most one failed `connect()`; the listen side cannot be the fix because a Bun client also has to reach non-Bun servers that bound `[::1]`, and usockets listens on a single fd; dropping `AI_ADDRCONFIG` for every name would bring back AAAA queries and unroutable IPv6 attempts on IPv4-only hosts, which is exactly the container case. `[::1, 127.0.0.1]` is also what glibc already returns for `localhost` on any machine with a link-local IPv6 address, so containers start behaving like everywhere else.

`node:net` / `node:dns` are deliberately unchanged (they mirror Node's `ADDRCONFIG` behaviour, bug included). The macOS dns_sd path is unchanged; localhost names are answered before it. The three hand-rolled addrinfo packers (`dns_sd_complete`, the test seeding hook, and now the synthetic answer) share one `addrinfo_chain`.

### Behavior changes and costs

- **`*.localhost` now resolves everywhere on this path.** `fetch("http://app.localhost:3000")`, `WebSocket`, `Bun.connect` and `bun install` reach loopback on macOS, Windows and musl too, where the OS resolver returns NXDOMAIN for `foo.localhost` today. This matches Chrome, Firefox and curl ≥ 7.85. An `/etc/hosts` entry that points `localhost` or `foo.localhost` at a non-loopback address is ignored by these APIs (still honored by `node:net` / `node:dns`).
- **Names that resolve only to loopback cost one extra lookup per cache TTL, on every host.** Rule 2's predicate is true for any single-family loopback answer, so a hosts-file dev alias (`127.0.0.1 myapp.test`) gets a second files-only `getaddrinfo` (microseconds), and a public name that resolves to `127.0.0.1` (`lvh.me`, `localtest.me`) gets a second DNS round trip including AAAA. Chrome pays the same cost. Every other name — anything with a non-loopback address, or loopback of both families — is untouched: one lookup, `AI_ADDRCONFIG` still set (pinned by the rule-2 table test).
- **No change to reported errors.** When every candidate fails, usockets reports `ECONNREFUSED` regardless of which address failed last (`context.c`), so adding `::1` on a host without IPv6 does not change the error a caller sees.

### Verification

- `test/js/bun/dns/dns-interleave.test.ts`: table tests for both predicates via two `bun:internal-for-testing` hooks (`dnsIsLocalhostName`, `dnsIsAllLoopbackOfOneFamily` — the only way to pin the negative cases like `localhost.example` without network), plus end-to-end tests that bind listeners on `127.0.0.1` and `::1` and connect through `localhost`, `localhost.`, `bun-dns-test.localhost`, `bun-dns-test.localhost.`, and through `ip6-localhost` / `ip6-loopback` / `localhost6` to whatever the system maps them to, each with `fetch()`, `WebSocket` and `Bun.connect()`. An address the kernel refuses to bind is skipped so the file runs on hosts without IPv6. All six e2e tests fail on 1.4.0 in the container above and pass on this branch.
- Now passing in that container: `fetch-preconnect.test.ts` 14/14, `bun-serve-html-entry.test.ts` 7/7, bun-install "--registry CLI flag", `dns-prefetch.test.ts`, `serve.test.ts`, `node-http.test.ts` (remaining failures there and in `socket.test.ts` / `bun-connect-x509.test.ts` are the `example.com` / proxy / root-port tests, which fail identically on 1.4.0: no network in the container).
- `cargo check -p bun_runtime` for `aarch64-apple-darwin`, `x86_64-pc-windows-msvc`, `x86_64-unknown-linux-musl`; clippy clean; `test/internal/source-lints` pass.

Supersedes #37442.

<details>
<summary>Resolver and Node output from the container (only IPv6 address is ::1 on lo; Debian /etc/hosts)</summary>

```
$ python3 -c 'import socket; ...'
localhost        plain      : [('AF_INET6', '::1'), ('AF_INET', '127.0.0.1')]
localhost        ADDRCONFIG : [('AF_INET', '127.0.0.1'), ('AF_INET', '127.0.0.1')]
ip6-localhost    plain      : [('AF_INET6', '::1')]
ip6-localhost    ADDRCONFIG : [('AF_INET', '127.0.0.1')]
localhost.       plain      : gaierror errno=-3 (Temporary failure in name resolution)
foo.localhost    plain      : gaierror errno=-3 (Temporary failure in name resolution)

$ node node_repro.mjs   # v26.3.0
node server bound: ::1 46503
dns.lookup(localhost, all, ADDRCONFIG): [ { address: '127.0.0.1', family: 4 }, { address: '127.0.0.1', family: 4 } ]
node http.get via localhost FAILED: ECONNREFUSED connect ECONNREFUSED 127.0.0.1:46503
node fetch via localhost FAILED: ECONNREFUSED connect ECONNREFUSED 127.0.0.1:46503
node net.connect FAILED: ECONNREFUSED

$ bun repro.ts   # released 1.4.0
server bound to localhost 35847
fetch via localhost FAILED: ConnectionRefused
Bun.connect via localhost: connect FAILED: ECONNREFUSED
```

</details>

<details>
<summary>Background: AI_ADDRCONFIG and the connect-path resolver</summary>

- `AI_ADDRCONFIG` is a `getaddrinfo` hint: return IPv6 addresses only if the host has an IPv6 address, and likewise for IPv4, so an IPv4-only host is not handed unroutable AAAA records. glibc counts only non-loopback interface addresses, so a family that exists only on `lo` is treated as absent, and a hosts-file `::1` entry comes back as `127.0.0.1`.
- The connect-path resolver is the process-wide cache in `mod internal` of `dns.rs`. usockets asks it for every `fetch()`, `WebSocket`, `Bun.connect()` and `bun install` connection (the QUIC client too); one `Request` per hostname is completed through `after_result_entries`, and usockets reads the packed `ResultEntry` list and opens up to four connection attempts at once, moving on from addresses that fail. `node:net` resolves through `node:dns` first and does not use it.
- RFC 6761 §6.3 reserves `localhost` and every name under it for the loopback addresses; a resolver may answer them without consulting anything. Chrome implements that as `ServeLocalhost` / `IsLocalHostname` (`net/dns/host_resolver_manager.cc`, `net/base/url_util.cc`) and keeps the `IsAllLocalhostOfOneFamily` retry for hosts-file loopback names in `net/dns/host_resolver_system_task.cc`.

</details>

<details>
<summary>Earlier revision of this PR</summary>

The first revision cleared `AI_ADDRCONFIG` for `localhost` / `*.localhost` and still asked `getaddrinfo`. Review asked for the two Chrome mechanisms above instead: the name check still depended on `/etc/hosts`, missed `localhost.`, and did nothing for hosts-file names such as `ip6-localhost`. That revision also passed the hostname into `get_hints()` on the macOS dns_sd path, which is no longer needed.

</details>

---

**no test proof** · iteration 0 · Platform-specific test(s) that do not run on this machine. Deferring to CI, which covers all platforms: test/js/bun/dns/dns-interleave.test.ts

### CODE REVIEW

**Comment by robobun** [comment]
Status: fix pushed, waiting on CI.

Reproduced on a Linux host whose only IPv6 address is `::1` (IPv4-only Docker network): `Bun.serve({ hostname: "localhost" })` binds `[::1]`, and `fetch("http://localhost:PORT/")`, `new WebSocket(...)` and `Bun.connect(...)` to the name all get connection refused because the connect-path resolver's `AI_ADDRCONFIG` lookup returns `127.0.0.1` only. `test/cli/install/bun-install.test.ts` "should support --registry CLI flag" fails on such a host for the same reason.

- Fail before: `USE_SYSTEM_BUN=1 bun test test/js/bun/dns/dns-interleave.test.ts -t "loopback names"` (the `::1` entry reports `ConnectionRefused` / `closed: 1006` / `ECONNREFUSED`)
- Pass after: `bun bd test test/js/bun/dns/dns-interleave.test.ts` (6 pass), plus the two bun-install tests above

**Comment by robobun** [comment]
Re the review above: agreed that the approach is the thing to sign off on. The three options (exempt loopback names on the connect side, change what the listen side binds, or drop `AI_ADDRCONFIG` altogether) are compared in the Fix section of the description; the short version is that the connect path already tries every address it is given, the listen side has to stay whatever the system says `localhost` is (a Bun client also has to reach non-Bun servers that bound `::1`), and dropping the flag for every name would change DNS traffic for public hostnames, so the exemption is limited to the names that only ever resolve to loopback.

`localhost.` (trailing dot) is intentionally left out so the rule stays identical to the existing `localhost` / `*.localhost` checks in `DevServer.rs` and `normalize_dns_name`.

No code changes from this review; waiting on CI.

**Comment by robobun** [comment]
<div><sup>Updated 9:05 PM PT - Aug 14th, 2026</sup></div>

:x: @robobun, your commit 8b34ba175057b6709b96919286f72e71ae6230f8 has some failures in [`Build #97117`](https://buildkite.com/bun/bun/builds/97117) ([All Failures](http://bun-linux2:6786/?commit=8b34ba175057b6709b96919286f72e71ae6230f8))


<hr />


🧪 &nbsp; To try this PR locally:

```zig
bunx bun-pr 38818
```

That installs a local version of the PR into your `bun-38818` executable, so you can run:

```bash
bun-38818 --bun
```

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Right diagnosis, but the fix should match Chrome's resolver exactly rather than a name check that only clears `AI_ADDRCONFIG`. Our connect-path hints were copied from Chromium (`default_hints` links `host_resolver_proc.cc`); we never ported the two loopback safety nets that go with them. Please make this PR do precisely what Chrome does:

**1. `localhost` / `*.localhost` never hit the resolver** — `HostResolverManager::ServeLocalhost` → `ResolveLocalHostname` ([host_resolver_manager.cc](https://source.chromium.org/chromium/chromium/src/+/main:net/dns/host_resolver_manager.cc;l=333)) returns `[::1, 127.0.0.1]` synthetically. Name match is `IsLocalHostname` ([url_util.cc](https://source.chromium.org/chromium/chromium/src/+/main:net/base/url_util.cc;l=582)): strip one trailing `.`, then ASCII-case-insensitive `== "localhost"` or ends with `.localhost`. No `getaddrinfo`, no `/etc/hosts`, no `AI_ADDRCONFIG` question at all (RFC 6761 §6.3 / draft-west-let-localhost-be-localhost). This replaces `is_loopback_name` + "call getaddrinfo with the flag off"; the current PR still depends on `/etc/hosts` contents and misses `localhost.`.

**2. Keep `AI_ADDRCONFIG` for everything else, with Chrome's post-call retry** — [host_resolver_system_task.cc](https://source.chromium.org/chromium/chromium/src/+/main:net/dns/host_resolver_system_task.cc;l=627): if the flag was set and the result `IsAllLocalhostOfOneFamily()` (every address is loopback and they are all one family), clear `AI_ADDRCONFIG` and call `getaddrinfo` again (crbug 42058 / 49024). Put it next to the existing `EAI_NONAME` retry at `dns.rs:2692`. This is what fixes the general case (`ip6-localhost`, `ip6-loopback`, custom `/etc/hosts` loopback names) — the same bug the PR describes, not covered by a name check.

(Chrome's third mechanism, the `HaveOnlyLoopbackAddresses` probe that clears the flag globally, is not needed once 2 is in.)

Also:
- Drop the macOS `protocol_for_hints(get_hints(Some(host)))` change — with 1 in place `localhost` never reaches `lookup_dns_sd`, and the PR already notes mDNSResponder exempts loopback.
- Comment on `is_loopback_name` is 8 lines; keep it to one line or none.
- Tests: add a case for `localhost.` and one where a non-`localhost` name resolves to loopback of one family via `/etc/hosts` (or at least the `ip6-localhost` alias present in Docker) exercising the retry.

`node:net` / `node:dns` stay as they are — they mirror Node's `net.js` `ADDRCONFIG` line exactly and Node has this same bug; only the internal connect-path resolver changes.

File: src/runtime/dns_jsc/dns.rs [hunk 1]
```diff
@@ -2600,6 +2668,32 @@ pub mod internal {
         results
     }
 
+    /// A getaddrinfo()-shaped list over `addrs` (filled-in `sockaddr_in`/`sockaddr_in6`) for
+    /// [`process_results`]; the nodes point into `addrs`, so it must outlive them.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/dns_jsc/dns.rs]: Cut to one line in 721cfc2.

File: src/runtime/dns_jsc/dns.rs [hunk 2]
```diff
@@ -2688,10 +2782,23 @@ pub mod internal {
                 .unwrap_or(ptr::null());
             let mut err = libc::getaddrinfo(host_ptr, service, &raw const hints, &raw mut addrinfo);
 
-            // optional fallback
-            if err == netc::EAI_NONAME && (hints.ai_flags & netc::AI_ADDRCONFIG) != 0 {
+            // AI_ADDRCONFIG judges a family by the non-loopback interfaces, so retry without it
+            // when it left nothing, or only the loopback addresses of one family (Chrome does both).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/dns_jsc/dns.rs]: Cut to one line in 721cfc2.

File: src/runtime/dns_jsc/dns.rs [hunk 3]
```diff
@@ -2424,6 +2425,73 @@ pub mod internal {
...
+    fn is_localhost_name(host: &[u8]) -> bool {
+        const DOT_LOCALHOST: &[u8] = b".localhost";
+        let host = host.strip_suffix(b".").unwrap_or(host);
+        strings::eql_case_insensitive_ascii(host, b"localhost", true)
+            || (host.len() >= DOT_LOCALHOST.len()
+                && strings::eql_case_insensitive_ascii(
+                    &host[host.len() - DOT_LOCALHOST.len()..],
+                    DOT_LOCALHOST,
+                    true,
+                ))
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/dns_jsc/dns.rs:2360]: Added `".localhost": true` and `".localhost.": true` to the table in 721cfc2; the `>=` is deliberate so the predicate matches Chrome's EndsWith.

**Comment by robobun** [comment]
@Jarred-Sumner reworked as requested, in 9788c37 (plus 721cfc2 for two bot nits):

1. `localhost` / `*.localhost` never reach the resolver: `is_localhost_name` is Chrome's `IsLocalHostname` (one trailing dot stripped, ASCII case-insensitive `localhost` or `.localhost` suffix), and `internal::getaddrinfo` completes those names with `[::1, 127.0.0.1]` (`localhost_results`) through the normal cache entry and `after_result_entries`, so usockets sees an ordinary completed lookup. No `getaddrinfo`, no `/etc/hosts`.
2. `AI_ADDRCONFIG` stays on for everything else. The `EAI_NONAME` retry in `work_pool_callback` now also fires when the call succeeded but the answer `is_all_loopback_of_one_family` (Chrome's predicate); the first answer is kept if the retry fails. This is what fixes `ip6-localhost` / `ip6-loopback` / `localhost6` and custom hosts-file loopback names.
3. The macOS `get_hints(Some(host))` change is gone; `get_hints()` has its old signature and `lookup_dns_sd` is untouched.
4. The long comment is gone; the remaining ones are one line each.
5. Tests: `localhost.`, `bun-dns-test.localhost` and `bun-dns-test.localhost.` are connected through end to end (listeners on `127.0.0.1` and on `::1`, each reached with `fetch()`, `WebSocket` and `Bun.connect()`), and `ip6-localhost` / `ip6-loopback` / `localhost6` are connected through to whatever the host's own hosts file maps them to, which exercises the retry on Debian and Docker hosts (skipped where the name is absent). Both predicates also have table tests via two `bun:internal-for-testing` hooks, since the negative cases (`localhost.example`, an ordinary one-family answer) cannot be shown end to end without network access. All of the end-to-end tests fail on the released build in the affected container and pass on this branch.

`node:net` / `node:dns` are unchanged. Description updated to describe the current shape; the first revision is summarized in a details block at the bottom.

**Comment by robobun** [comment]
The `reachThroughName` fixture from this PR can reach a listener of another process on the other loopback address with the same port number. #41448 makes the fixture hold the port on both addresses.
